src/edu/stanford/nlp/optimization/QNMinimizer.java

```diff
@@ -522,45 +521,55 @@ private eState toContinue(StringBuilder sb) {
     // Diagonal Options
     // Line search Options
     // Memory stuff
-    private List<double[]> s = null;
-    private List<double[]> y = null;
-    private List<Double> rho = null;
+    private double[][] s;
+    private double[][] y = null;
+    private double[] rho = null;
     private double gamma;
     public double[] d = null;
-    private int mem;
-    private int maxMem = 20;
+    private int mem = 20, used = 0;
     public eScaling scaleOpt = eScaling.SCALAR;
 
     QNInfo(int size) {
-      s = new ArrayList<>();
-      y = new ArrayList<>();
-      rho = new ArrayList<>();
+      mem = size > 0 ? size : 20;
+      s = new double[mem][];
+      y = new double[mem][];
+      rho = new double[mem];
       gamma = 1;
-      mem = size;
     }
 
     QNInfo(List<double[]> sList, List<double[]> yList) {
-      s = new ArrayList<>();
-      y = new ArrayList<>();
-      rho = new ArrayList<>();
+      s = new double[mem][];
+      y = new double[mem][];
+      rho = new double[mem];
       gamma = 1;
       setHistory(sList, yList);
     }
 
     int size() {
-      return s.size();
+      return used;
     }
 
     double getRho(int ind) {
-      return rho.get(ind);
+      return rho[ind];
     }
 
     double[] getS(int ind) {
-      return s.get(ind);
+      return s[ind];
     }
 
     double[] getY(int ind) {
-      return y.get(ind);
+      return y[ind];
+    }
+
+    void removeFirst() {
```

Comment:
Do you have any interest in doing the circular buffer?  Should I take a look at it?

Reply:
Actually, even among micro-optimizations this would be the most micro of optimizations, right?  It's basically copying 20 references per iteration, which would be the tiniest drop in the bucket of the processing time for model training

Reply:
I am quite busy right now, so I can't reply to all comments, and much of the edits I did were like 1.5-2 years ago.
A circular buffer may be more elegant: instead of having a `used` that will at some point always be 19 and 20, you would have a counter pointing to the next element to be replaced; and if you need shrinking at all, one could overwrite values with `null`. Right now, what is the meaning of the position of a value in this arrays except recency? Once filled, the current code removes s[0], y[0]; moves everything up front, and then reinserts s and y at the end again (because the code now recycles the arrays; it most likely did not do so originally). A simple "latest=-1" variable should suffice, and when you do a new solution, you increase latest (wrapping around to 0) and work on s[latest] and y[latest]. So it would likely result in simpler code.